iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -0,0 +1,378 @@
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Devices.Shared;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// A collection of properties for the client.
+    /// </summary>
+    public class ClientPropertyCollection : PayloadCollection
+    {
+        private const string VersionName = "$version";
+
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
+        /// </summary>
+        /// <remarks>
+        /// If the collection has a key that matches the property name this method will throw an <see cref="ArgumentException"/>.
+        /// <para>
+        /// When using this as part of the writable property flow to respond to a writable property update you should pass in the value
+        /// as an instance of <see cref="PayloadSerializer.CreateWritablePropertyResponse(object, int, long, string)"/>
+        /// to ensure the correct formatting is applied when the object is serialized.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="ArgumentNullException"><paramref name="propertyName"/> is <c>null</c> </exception>
+        /// <param name="propertyName">The name of the property to add.</param>
+        /// <param name="propertyValue">The value of the property to add.</param>
+        public override void Add(string propertyName, object propertyValue)
+            => Add(null, propertyName, propertyValue);
+
+        /// <inheritdoc path="/remarks" cref="Add(string, object)" />
+        /// <inheritdoc path="/seealso" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
+        /// </summary>
+        /// <exception cref="ArgumentNullException"><paramref name="propertyName"/> is <c>null</c> </exception>
+        /// <param name="componentName">The component with the property to add.</param>
+        /// <param name="propertyName">The name of the property to add.</param>
+        /// <param name="propertyValue">The value of the property to add.</param>
+        public void Add(string componentName, string propertyName, object propertyValue)
+            => AddInternal(new Dictionary<string, object> { { propertyName, propertyValue } }, componentName, false);
+
+        /// <inheritdoc path="/remarks" cref="Add(string, object)" />
+        /// <inheritdoc path="/seealso" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <inheritdoc path="/exception['ArgumentException']" cref="AddInternal(IDictionary{string, object}, string, bool)" />
+        /// <summary>
+        /// Adds the value for the collection.
```

Comment:
Copy paste error? Wouldn't this be "Adds the value**s**" or "Adds all the value**s**"?